Remove starting_items redundancy in plando/spoiler (part 1)

fenhl opened the pull request:
This addresses the first 2 steps of #1349:

* Allow the `starting_items` setting to be plando'd with the syntax currently used by the top-level `starting_items` entry
    * This will be in addition to the current syntax for the setting, which will continue to be used by the GUI for now
    * Checks are in place to prevent using the dict-style syntax for `starting_items` while also having `starting_equipment` and/or `starting_songs` present (even if the former is plando'd and the latter comes from the GUI), which doesn't make much sense since dict-style `starting_items` includes equipment and songs.
* Remove the top-level entry, replacing it with an informative (i.e. ignored in plando) `:skipped_locations` entry

I have tested to make sure the new syntax works as expected, and the GUI still works as it did before. This also fixes #1353, as well as the bug where a Triforce piece on free Song from Impa would overwrite plando'd starting pieces instead of being added.

fenhl committed:
Remove starting_items redundancy in plando/spoiler

fenhl committed:
Add ammo to starting items from skipped locations, fixes #1353

fenhl committed:
Remove unnecessary None check

Zannick commented on Plandomizer.py:
```diff
@@ -911,26 +907,46 @@ def configure_gossip(self, spoiler, stoneIDs):
+    def configure_effective_starting_items(self, worlds, world):
+        items = {item_name: record.copy() for item_name, record in self.starting_items.items()}
+
+        if world.start_with_rupees:
```
I think these need to be `world.setting.<setting>` now.

fenhl committed:
Adjust for world.settings refactor

Roman971 commented:
I commented with my opinions on the issue (#1349), which is related to this PR. I'm mostly concerned about the potential confusion introduced with this solution, tho I agree the current situation isn't ideal and we need some changes.

fenhl committed:
Replace :effective_starting_items in spoiler log with :skipped_locations

fenhl committed:
Include HC Malon Egg in skipped locations list

fenhl committed:
Update for upstream changes

This fixes the bingo hints fixes by making use of the new
effective_starting_items variable.

Also fixes dict-style starting_items being rejected by the plando
validator.

fenhl committed:
Remove top-level starting items handling

cjohnson57 commented:
Not sure if it's due to updates since you last worked on it, but this seems to break plando'd starting items. I was testing with this one:

```json
{
  "starting_items": {
    "Gold Skulltula Tokens": 47
  }
}
```

And it neither printed the tokens in the spoiler nor actually gave them to me in game.

fenhl commented:
This PR removes the `starting_items` top-level entry. Maybe there should be a check and error message for that as well?

cjohnson57 commented:
My bad, for some reason the way I read it it was the one in settings getting axed and them all being combined into that one.. Which is pretty dumb of me since the description mentions "top-level" multiple places

fenhl commented:
According to @cjohnson57, the following plando causes a false positive of the “Incompatible starting item settings” exception that is intended to catch mixing dict-style `starting_items` with `starting_equipment` or `starting_songs`. I'll need to investigate that, marking as a draft until it's addressed.

```json
{
  "settings": {
    "starting_items": {
      "Gold Skulltula Token": 47
    }
  }
}
```

fenhl commented:
I am unable to reproduce that issue.

fenhl committed:
Explicit error when trying to use top-level starting_items

Zannick commented on SaveContext.py:
```diff
@@ -267,6 +265,8 @@ def give_item(self, item, count=1):
             self.give_health(count)
         elif item == "Bombchu Item":
             self.give_bombchu_item()
+        elif item == "Recovery Heart": # ItemPool.IGNORE_LOCATION
```
I'd prefer importing the constant and using it here.

Zannick replied:
I will do this, no need to worry.

fenhl replied:
With that change, it might make sense to preserve the info that this represents a recovery heart, maybe by simply swapping the expression and the comment, to clarify that that item is being handled. The reason I didn't do it this way is that it would lead to an import cycle.
